### PR TITLE
[Merged by Bors] - Use SmallVec in Bitfield

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Alternative to #2935

## Proposed Changes

Replace the `Vec<u8>` inside `Bitfield` with a `SmallVec<[u8; 32>`. This eliminates heap allocations for attestation bitfields until we reach 500K validators, at which point we can consider increasing `SMALLVEC_LEN` to 40 or 48.

While running Lighthouse under `heaptrack` I found that SSZ encoding and decoding of bitfields corresponded to 22% of all allocations by count. I've confirmed that with this change applied those allocations disappear entirely.

## Additional Info

We can win another 8 bytes of space by using `smallvec`'s [`union` feature](https://docs.rs/smallvec/1.8.0/smallvec/#union), although I might leave that for a future PR because I don't know how experimental that feature is and whether it uses some spicy `unsafe` blocks.

### CODE REVIEW

**Comment by michaelsproul** [comment]
doc tests broken, will fix

**Comment by paulhauner** [comment]
Some rough napkin math shows an attestation will be 256 bytes after this:

```
32 # aggregation bits
8 # data.slot
8 # data.index
32 # data.beacon_block_root
8 # data.source.epoch
32 # data.source.root
8 # data.target.epoch
32 # data.target.root
96 # signature

32 + 8 + 8 + 32 + 8 + 32 + 8 + 32 + 96 = 256
```

That sounds like a pretty tidy number. Power of two and all :tada:

File: consensus/ssz_types/src/bitfield.rs
Comment (paulhauner) [inline consensus/ssz_types/src/bitfield.rs:1330]: I spent a while thinking on this one... Is this comparison used just because `BitList1024` is on-hand? :sweat_smile:

**Review by paulhauner — APPROVED** [review]
Looks good! Let's finally merge this :tada:

File: consensus/ssz_types/src/bitfield.rs
Comment (michaelsproul) [reply consensus/ssz_types/src/bitfield.rs:1330]: Yeah, and all BitLists should be the same size

**Comment by michaelsproul** [comment]
The bitfield will be 56 bytes rather than 32 bytes because of the 3 extra words for smallvec enum discriminant, smallvec length and bitfield length.

This throws off the 256 byte alignment a little, but we could maybe come back for that? Usually we store attestations in Vecs, so I think they can cross page boundaries in the Vec's contiguous allocation. E.g. for 128 attestations * 280 bytes we have 35840 bytes, which can fit in 9 pages rather than 8 pages if they were 256 bytes or less. I think right now Attestations *are* 256 bytes (because the bitfield is 24 vec bytes plus 8 length bytes).

Maybe let's run this on some nodes for a bit before merging? It definitely seemed to help on my tree-states branch but it would be good to confirm here

**Comment by michaelsproul** [comment]
Ah, so, attestations aren't anywhere close to 256 bytes because signatures aren't actually 96 bytes, they're 304 bytes :exploding_head:

Signatures correspond to `blst`'s [`blst_p2`](https://github.com/supranational/blst/blob/3491df1a0b745c7c73efc7ac9390a2265ff00b1d/bindings/rust/src/lib.rs#L1549) type, which is [288 bytes](https://github.com/supranational/blst/blob/3491df1a0b745c7c73efc7ac9390a2265ff00b1d/bindings/rust/src/bindings.rs#L574-L587) and we add 16 bytes for the `Option<AggSig>` and `is_infinity: bool` in `GenericAggregateSignature`.

Therefore this PR increases the size of attestations in memory from 472 bytes to 488 bytes, which doesn't break any nice alignment properties. I've added a test to regression test this size so that it doesn't change unexpectedly and we have a go-to reference for this value, in this commit https://github.com/sigp/lighthouse/pull/3025/commits/c7c4f39c6b3f2415ad0b339d7873e72b7c610340. The test will only pass with `blst`, but seeing as we don't run the `consensus/types` with Milagro I think that's an acceptable compromise. I've also used some hard-coded constants so that it's easy to read.

**Review by paulhauner — APPROVED** [review]
My math was way off! Thanks for correcting it.

This looks good. I'm not certain why we need the `const_generics` feature, but I'll leave that up to you.

**Comment by michaelsproul** [comment]
> This looks good. I'm not certain why we need the const_generics feature, but I'll leave that up to you.

Good catch! I've gotten rid of it which might save us some compile time. We can re-enable in future if we want `SMALLVEC_LEN = 40`.

I'm running this on Frankfurt atm and it looks good, but it hasn't really been long enough for fragmentation to occur. I'll let it run overnight and then merge in the morning (atop #3024). Then I'll keep an eye on our `unstable` derived builds for a while to see if the fragmentation situation seems any better.

**Comment by michaelsproul** [comment]
Frankfurt looking good after the most recent restart, although I think it's still a bit too early to say if it'll revert to its old behaviour.

frankfurt

**Pull request closed without merging**
